gossip/src/duplicate_shred.rs

```diff
@@ -118,6 +114,31 @@ where
         }
     }
 
+    // Merkle root conflict check
+    if shred1.fec_set_index() == shred2.fec_set_index() {
+        let mr1 = shred1.merkle_root().ok();
+        let mr2 = shred2.merkle_root().ok();
+
+        if (mr1.is_some() && mr2.is_none()) || (mr2.is_some() && mr1.is_none()) {
+            // A mixture of legacy and merkle shreds in the same fec set
+            // constitutes a valid duplicate proof
+            return Ok(());
+        }
+
+        if let Some((mr1, mr2)) = mr1.zip(mr2) {
+            if mr1 != mr2 {
+                // Conflicting merkle roots for the same fec set is a
+                // valid duplicate proof
+                return Ok(());
+            }
+        }
+    }
+
+    if shred1.shred_type() != shred2.shred_type() {
+        // Only valid proof here is a merkle conflict which was checked above
+        return Err(Error::InvalidMerkleRootConflict);
```

Comment:
The return error value is going to be very confusing if not wrong when other types of duplicates mentioned in https://github.com/solana-labs/solana/pull/33979#issuecomment-1802559831 are included here, because those also can be different shred-type but are not merkle-root conflict.

Can we keep the error `ShredTypeMismatch`? because here we are checking for shreds type to be matching and for the rest of this function we expect the shreds types to be the same.